### PR TITLE
[ML] Report thread settings per node for trained model deployments

### LINKED ISSUE #81149: [ML] Allocation stats should report actual thread settings used
PyTorch trained model deployments have 2 thread settings `inferenceThreads` and `modelThreads` which determine how compute resources are used. To prevent over-subscription the native process [imposes a limit on the settings](https://github.com/elastic/ml-cpp/blob/main/bin/pytorch_inference/Main.cc#L211) which are checked against `std::thread::hardware_concurrency()`. Any adjustments to the settings are logged but those log messages are not the most visible, it would be better to report back in the deployment stats that a setting was modified. This should be done per node as models are deployed across heterogenous nodes which can have different hardware profiles and values of `std::thread::hardware_concurrency()`.

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
When a trained model deployment is started the user may request specific
number of threads for inference and/or parallel forwarding. However,
when the pytorch process starts these settings are adjusted so that
the actual number of threads doesn't exceed the available number of
threads. This commit changes the result format so that other type of
results may be read from the process than just inference results.
We then introduce a result for reading the thread settings from the
process and we report it in the response of the trained model _stats API.

Closes #81149

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/process/PyTorchResultProcessor.java [hunk 1]
```diff
@@ -85,23 +90,30 @@ public void process(NativePyTorchProcess process) {
...
-        return new LongSummaryStatistics(timingStats.getCount(), timingStats.getMin(), timingStats.getMax(), timingStats.getSum());
-    }
-
-    private synchronized void processResult(PyTorchResult result) {
-        if (result.isError() == false) {
-            timingStats.accept(result.getTimeMs());
+    private void processInferenceResult(PyTorchInferenceResult inferenceResult) {
+        logger.trace(() -> new ParameterizedMessage("[{}] Parsed result with id [{}]", deploymentId, inferenceResult.getRequestId()));
+        if (inferenceResult.isError() == false) {
+            timingStats.accept(inferenceResult.getTimeMs());
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/process/PyTorchResultProcessor.java:104]: `LongSummaryStatistics` is not thread safe which is why this method was previously synchronized. 

Perhaps extract the line `timingStats.accept(inferenceResult.getTimeMs());` into a separate synchronized method

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/process/PyTorchResultProcessor.java [hunk 2]
```diff
@@ -55,13 +60,13 @@ public void process(NativePyTorchProcess process) {
...
-                logger.trace(() -> new ParameterizedMessage("[{}] Parsed result with id [{}]", deploymentId, result.getRequestId()));
-                processResult(result);
-                PendingResult pendingResult = pendingResults.remove(result.getRequestId());
-                if (pendingResult == null) {
-                    logger.debug(() -> new ParameterizedMessage("[{}] no pending result for [{}]", deploymentId, result.getRequestId()));
-                } else {
-                    pendingResult.listener.onResponse(result);
+                PyTorchInferenceResult inferenceResult = result.inferenceResult();
+                if (inferenceResult != null) {
+                    processInferenceResult(inferenceResult);
```
[the remark this answers is not in the document]
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/process/PyTorchResultProcessor.java:65]: It doesn't hurt to not have it like this though. I know we'll probably not be compounding results together, but we follow this pattern for autodetect and analytics. I'd rather we kept it like this here too.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/process/PyTorchResultProcessor.java (hunk 1 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/pytorch/process/PyTorchResultProcessor.java]: Ah, well spotted. I added those two lines in a `synchronized` block.

**Review by davidkyle — APPROVED** [review]
LGTM

**Pull request merged**